codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+            streamInfo.childChannel.fireChildReadComplete();
+        }
+        streamsToFireChildReadComplete.clear();
+    }
+
+    void fireChildReadAndRegister(StreamInfo streamInfo, Object msg) {
+        // Can't use childChannel.fireChannelRead() as it would fire independent of whether
+        // channel.read() had been called.
+        streamInfo.childChannel.fireChildRead(msg);
+        if (!streamInfo.inStreamsToFireChildReadComplete) {
+            streamsToFireChildReadComplete.add(streamInfo);
+            streamInfo.inStreamsToFireChildReadComplete = true;
+        }
+    }
+
+    final class ConnectionListener extends Http2ConnectionAdapter {
+        @Override
+        public void onStreamActive(Http2Stream stream) {
+            if (ctx == null) {
+                // UPGRADE stream is active before handlerAdded().
+                return;
+            }
+            // If it is an outgoing stream, then we already created the channel.
+            // TODO: support outgoing streams
+            if (stream.getProperty(streamInfoKey) != null) {
+                return;
+            }
+            stream.setProperty(streamInfoKey,
+                new StreamInfo(createStreamChannel(ctx, stream, streamHandler)));
+        }
+
+        @Override
+        public void onStreamClosed(Http2Stream stream) {
+            final StreamInfo streamInfo = (StreamInfo) stream.removeProperty(streamInfoKey);
+            if (streamInfo != null) {
+                if (streamInfo.childChannel.eventLoop().inEventLoop()) {
+                    onStreamClosed0(streamInfo);
+                } else {
+                    streamInfo.childChannel.eventLoop().execute(new OneTimeTask() {
+                        @Override
+                        public void run() {
+                            onStreamClosed0(streamInfo);
+                        }
+                    });
+                }
+            }
+        }
+
+        private void onStreamClosed0(StreamInfo streamInfo) {
+            streamInfo.childChannel.onStreamClosedFired = true;
+            streamInfo.childChannel.fireChildRead(AbstractHttp2StreamChannel.CLOSE_MESSAGE);
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(errorCode, debugData);
+            try {
+                http2Handler.connection().forEachActiveStream(new Http2StreamVisitor() {
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                && http2Handler.connection().local().isValidStreamId(stream.id())) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                            streamInfo.childChannel.pipeline()
+                                    .fireUserEventTriggered(goAway.retain());
```

Comment:
Hmm... You're right. That's unfortunate, but it does seem we will need to duplicate each. I've added a TODO to try to figure out an alternative.